### ISSUE TITLE
default_factory doesn't work

### ISSUE THREAD

**Issue by ShtykovaAA** [strawberry-graphql/strawberry#3517]
Hi! 
I use default_factory to initialize my variable, but variable always returns the same result. It seems like default_factory doesn't work and it returns always the same result of function.

Here is example to reproduce:
https://play.strawberry.rocks/?gist=a7a5e62ffe4e68696b44456398d11104

**Comment by patrick91** [comment]
I can reproduce this, I think it's because we store the default value when creating the object, see: 

![CleanShot 2024-05-28 at 16 35 47@2x](https://github.com/strawberry-graphql/strawberry/assets/667029/a0952327-c789-4afa-ba48-13325df76c7e)

**Comment by coady** [comment]
I don't think this can be supported (and still be compliant with the spec). GraphQL defaults are statically in the schema. `strawberry export-schema ...`:
```graphql
type Mutation {
  update1(fields: MyInputType!): String!
  update2(fields: MyInputType!): String!
}

input MyInputType {
  field1: String = "318fbf6e-73b6-40eb-932f-0b66ba935b75"
}

type Query {
  hello: String!
}
```

Another issue is that the client sending an explicit null is valid and [semantically different](http://spec.graphql.org/October2021/#sec-Null-Value). So `MyInputType` would need logic like
```python
if field1 in (None, UNSET):
    field1 = uuid_pkg.uuid4()
```
That may seem like a workaround, but is actually the only correct implementation.

**Comment by ShtykovaAA** [comment]
@patrick91 and also can I ask in what situations then we need to use `default_factory` if this field is static, when we can use only `default`?

**Comment by patrick91** [comment]
> @patrick91 and also can I ask in what situations then we need to use `default_factory` if this field is static, when we can use only `default`?

I'm not sure to be honest, I'll need to think about this a bit

I do think it might be a flaw, or at least something surprising, so maybe we need to reconsider it

**Comment by bellini666** [comment]
I do understand the schema's default value issue, but I do agree with [this comment](https://github.com/strawberry-graphql/strawberry/issues/3517#issuecomment-2136089398). I actually [had to do workaround a similar issue on strawberry-resources](https://github.com/blb-ventures/strawberry-resources/blob/main/strawberry_resources/integrations/django.py#L191) when exporting form data for the field as a dynamic default value would not actually make sense there.

So my vote would be to actually change the behavior to fix this issue, specially since we still are 0.x =P, and mention as a "possible breaking change" in the changelog, mentioning the use of `default` as the correct way of relying on the older behavior

**Comment by patrick91** [comment]
just throwing out some ideas to make the change less painful

1. we could a `static_factory` or `schema_factory` which will have the current behaviour, `default_factory` will mimic dataclasses' and pydantic's behaviour
2. have a configuration option to disable static defaults
3. just change the behaviour

**Comment by patrick91** [comment]
@coady sorry to ping you again, but do you use default_factory for defaults in the schema level? 😊 what's your use case exactly?

**Comment by coady** [comment]
> @coady sorry to ping you again, but do you use default_factory for defaults in the schema level? 😊 what's your use case exactly?

The only use I'm aware of (and use) is for mutables, as [dataclasses requires](https://docs.python.org/3/library/dataclasses.html#dataclasses.field). Any valid value is a valid default value, including `[]` and `{}`.

I'd be happy with a cleaner alternative for mutables. This is forbidden (but is valid GraphQL):
```python
    q: list[float] = [0.5]
```
So instead I have to used `default_factory` or this:
```python
    q: list[float] = (0.5,)  # type: ignore
```
which mypy complains about.

**Comment by patrick91** [comment]
> just throwing out some ideas to make the change less painful
> 
> 1. we could a `static_factory` or `schema_factory` which will have the current behaviour, `default_factory` will mimic dataclasses' and pydantic's behaviour
> 2. have a configuration option to disable static defaults
> 3. just change the behaviour
> 
> option 1 could also have a codemod to make the update easier 😊

I think we could do option 2 for the time being
